modules/data_release/ajax/AddPermission.php

```diff
@@ -12,25 +12,92 @@
+             'data_release_id' => $data_release_id,
+            )
+        );
+    } elseif (empty($_POST['data_release_id'])
+        && !empty($_POST['data_release_version'])
+    ) {
+        $userid = $_POST['userid'];
+        $data_release_version = $_POST['data_release_version'];
 
-    $factory  = NDB_Factory::singleton();
-    $settings = $factory->settings();
+        $IDs = $DB->pselectCol(
+            "SELECT id FROM data_release WHERE "
+            . "version=:data_release_version",
+            array('data_release_version' => $data_release_version)
+        );
 
-    $baseURL = $settings->getBaseURL();
+        foreach ($IDs as $ID) {
+            $success = $DB->insert(
+                'data_release_permissions',
+                array(
+                 'userid'          => $userid,
+                 'data_release_id' => $ID['id'],
+                )
+            );
+        }
+    }
 
     header("Location: {$baseURL}/data_release/?addpermissionSuccess=true");
+} elseif ($_POST['action'] == 'managepermissions'
+    && $user->hasPermission('superuser')
+) {
+    // 1) so with checkboxes it's not straightforward to figure out
+    //    which permissions were either set or unset during the save,
+    //    hence the TRUNCATE at the beginning
+    // 2) put it in a transaction to make sure it succeeds or not at all
+    // 3) some special characters become "_" so we wildcard match the username
+    try {
+        $DB->_PDO->beginTransaction();
+        $DB->run("TRUNCATE data_release_permissions");
+        foreach ($_POST as $key => $value) {
+            if (strpos($key, 'permissions') !== false) {
+                $parsed_user = str_replace(
+                    "_",
+                    "%",
+                    str_replace("permissions_", "", $key)
+                );
+                $userid      = $DB->pselectOne(
+                    "SELECT ID FROM users WHERE UserID LIKE :userid",
+                    array('userid' => $parsed_user)
+                );
+                foreach ($value as $k => $v) {
+                    $data_release_ids = $DB->pselect(
+                        "SELECT id FROM data_release WHERE version=:version",
+                        array('version' => $v)
+                    );
+                    foreach ($data_release_ids as $data_release_id) {
+                        $success = $DB->run(
+                            "INSERT IGNORE INTO data_release_permissions VALUES "
+                            . "($userid, {$data_release_id['id']})"
+                        );
+                    }
+                }
+            }
+        }
+        $DB->_PDO->commit();
+        header("HTTP/1.1 303 See Other");
+        header("Location: {$baseURL}/data_release/?addpermissionSuccess=true");
+    } catch (Exception $e) {
+        $DB->_PDO->rollback();
+        error_log("ERROR: Did not update Data Release permissions.");
+        header("HTTP/1.1 500 Internal Server Error");
```

Comment:
Not a big deal for this PR but in general it might be nice not to hardcode the HTTP version and instead use `$_SERVER['SERVER_PROTOCOL']` or simply [`http_response_code`](https://secure.php.net/manual/en/function.http-response-code.php) instead.

HTTP/2 is around now and I'm not sure how sending raw HTTP headers might affect the connection -- but would be a shame to downgrade to a slower protocol by mistake.

Reply:
I think this can be a general LORIS-wide improvement at some point